1-Draft/RFCXXXX-RFC-Invoke-DscResource.md

```diff
@@ -0,0 +1,246 @@
+implement [least-privilege Administrative Models](https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/plan/security-best-practices/implementing-least-privilege-administrative-models),
+it will **not** be supported as part of this implementation of `Invoke-DscResource`.
+
+Instead, we'll aim at providing an example function in a script published
+on the PowerShell Gallery to proxy this command, so that it handles `PsDscRunAsCredential`
+to invoke the resource in a different user context, using PowerShell Jobs.
+
+If the key `PsDscRunAsCredential` is to be found amongst the keys of the
+`-Property` parameter, then `Invoke-DscResource` will throw an exception,
+unless the `Invoke-DscResource` is invoked with the (new) switch parameter `-IgnorePsDscRunAsCredential`,
+in which case it will invoke the DSC Resource method after stripping
+the `PsDscRunAsCredential` key/value pair from the properties.
+
+So calling:
+
+```PowerShell
+Invoke-DscResource -Name Script -ModuleName @{ModuleName='PSDscResources';ModuleVersion='2.12.0.0'} -Method 'Set' -Property @{
+    GetScript  = '<# My Get ScriptBlock #>'
+    SetScript  = '<# My Set ScriptBlock #>'
+    TestScript = '<# My Test ScriptBlock #>'
+    PsDscRunAsCredential = $Credential
+} -Verbose
+
+```
+
+Will **throw an exception**, because of the `PsDscRunAsCredential` key.
+
+> NOTES: We're avoiding to take dependencies on other technologies that would require extra configurations or permissions (such as remoting), or would not be available on other OSes.
+
+#### Independent and isolated execution
+
+`Invoke-DscResource` in PowerShell 7+ will not be aware of other instances being executed, and as such it will be possible to execute several instances in parallel when isolated in their own runspace, or run in parallel with the LCM.
+
+This means that it enables concurrent execution, but also risks conflict if two conflicting resources are run simultaneously.
+
+It is up to the user to sequence the execution safely, or to create appropriate resources.
+
+#### Output & Types
+
+##### Get
+
+The `GET` function invoked via the LCM (WMF 5.1) returns a CIM representation of a hashtable,
+and will be a `hashtable` for `Invoke-DscResource` in PS7+.
+
+##### Test
+
+The `TEST` function in WMF 5.1 returns a CIM object, with a `[bool]` NoteProperty
+`InDesiredState` that has the boolean result of the test.
+
+```PowerShell
+InDesiredState
+--------------
+True
+```
+
+In PS7+, `Invoke-DscResource -Method Test ...` will return an object (not CIM)
+that has the same `InDesiredState` Property.
+
+##### Set
+
+The `SET` function in WMF 5.1 returns a CIM object, with a `[bool]` NoteProperty
+`RebootRequired`, corresponding whether the `$global:DSCMachineStatus` has been
+set to 1.
+
+```PowerShell
+RebootRequired
+--------------
+False
+```
+
+In PS7+, `Invoke-DscResource -Method Set ...` will return an object (not CIM) that
+has the same `RebootRequired` Property, based on whether `$global:DSCMachineStatus`
```

Comment:
Given that this will now run in the current user's current scope, how safe is `$global:DSCMachineStatus`?  Will there be any validation of the presence or absence of that variable and it's state before the execution of `Invoke-DscResource`?

Reply:
My implementation is resetting the state before running the resource.  I don't see any other safe way to do this.

Reply:
I was actually wondering if we need to use it at all for `Invoke-DscResource`.
Maybe we could reset it to what it is when the function is called, then only return the need to reboot in the return object (`RebootRequired`)?

Or we after we reset before execution, we could do a `$Before -bor $After`?
What do you think?

Reply:
Be have no idea who set it in the before state.  This is thinking like we are in the LCM engine.  The CmdLet should not be vulnerable to manipulation by global variable affected about side the cmdlet.  This is a non-started from a security perspective.

Reply:
As discussed offline with @TravisEz13, we'd rather reset the variable at the beginning (because we can't trust it), and the user should not use that global variable after either.

Instead, the user should use the `RebootRequired` property of the object returned by the `Invoke-DscResource` cmdlet when executing the `Set` method.

To avoid misuse of `$global:DSCMachineStatus` after the resource is invoked, we'd rather remove it completely from the session.

I will update the RFC accordingly.